baseapp/baseapp.go

```diff
@@ -175,13 +189,65 @@ func (app *BaseApp) MountStore(key sdk.StoreKey, typ sdk.StoreType) {
 // LoadLatestVersion loads the latest application version. It will panic if
 // called more than once on a running BaseApp.
 func (app *BaseApp) LoadLatestVersion(baseKey *sdk.KVStoreKey) error {
-	err := app.cms.LoadLatestVersion()
+	err := app.storeLoader(app.cms)
 	if err != nil {
 		return err
 	}
 	return app.initFromMainStore(baseKey)
 }
 
+// DefaultStoreLoader will be used by default and loads the latest version
+func DefaultStoreLoader(ms sdk.CommitMultiStore) error {
+	return ms.LoadLatestVersion()
+}
+
+// StoreLoaderWithUpgrade is used to prepare baseapp with a fixed StoreLoader
+// pattern. This is useful in test cases, or with custom upgrade loading logic.
+func StoreLoaderWithUpgrade(upgrades *storeTypes.StoreUpgrades) StoreLoader {
+	return func(ms sdk.CommitMultiStore) error {
+		return ms.LoadLatestVersionAndUpgrade(upgrades)
+	}
+}
+
+// UpgradeableStoreLoader can be configured by SetStoreLoader() to check for the
+// existence of a given upgrade file - json encoded StoreUpgrades data.
+// If the file if present, parse it and execute those upgrades
+// (rename or delete stores), while loading the data.
+//
+// This is useful for in place migrations when a store key is renamed between
+// two versions of the software.
+func UpgradeableStoreLoader(upgradeInfoPath string) StoreLoader {
+	return func(ms sdk.CommitMultiStore) error {
+		_, err := os.Stat(upgradeInfoPath)
+		if os.IsNotExist(err) {
+			return DefaultStoreLoader(ms)
+		} else if err != nil {
+			return err
+		}
+
+		// there is a migration file, let's execute
+		data, err := ioutil.ReadFile(upgradeInfoPath)
+		if err != nil {
+			return fmt.Errorf("Cannot read upgrade file %s: %v", upgradeInfoPath, err)
+		}
+		var upgrades storeTypes.StoreUpgrades
+		err = json.Unmarshal(data, &upgrades)
+		if err != nil {
+			return fmt.Errorf("Cannot parse upgrade file: %v", err)
+		}
+		err = ms.LoadLatestVersionAndUpgrade(&upgrades)
+		if err != nil {
+			return fmt.Errorf("Load and upgrade database: %v", err)
+		}
+		// if we have a successful load, we delete the file
```

Comment:
godoc doesn't state this either. Should we even do this?

Reply:
I agree the godoc should be upgraded to mention this. I wonder if the idea is that it should be deleted because the data has now been upgraded? Seems odd though - I wish there was a way to instead just check if the upgrades are relevant and then upgrade if they are. This way the upgrade file wouldn't need to be eradicated

Reply:
I will update the godocs.

This is written for the workflow from the regen network upgrade package. They need to run the upgrade store *exactly once* on first restart. Since we cannot load the db before renaming stores, this upgrade info was to be stored on disk.

I agree this code doesn't really belong in baseapp. I would move it to the upgrade module when that #4233 PR is finally merged. Are you okay with leaving this in for now, and then moving it to the upgrade module later? Or should I just remove this now?

Reply:
General comment:

There is a big feature that we want to move into the sdk, but seems to take a while, so I am trying to break out lots of smaller PRs for it. They are all functional, but some may not actually fill a user story without the others. 

In this case, the code is meant to be used by another PR with packages not present here. What is the best way to go about this in general?

Reply:
> I agree this code doesn't really belong in baseapp. I would move it to the upgrade module when that #4233 PR is finally merged. Are you okay with leaving this in for now, and then moving it to the upgrade module later? Or should I just remove this now?

Mmmhhh, leave as-is for now I suppose. As long as we follow up in #4233.


> In this case, the code is meant to be used by another PR with packages not present here. What is the best way to go about this in general?

I would create an issue to act as an epic (in agile lingo). This issue/epic lays out all the changes desired and their motive. Then also have a checklist referencing each PR.

Reply:
There is now a TODO to mark that.

Part of the issue IMO is that upgrade module PR is waiting since months and thus I don't want to add any more functionality to it (to be blocked), so as I worked to add support for a v0.34 -> v0.36 inplace upgrade, I had to make separate PRs to address them, as the root PR seems to be on indefinite hold.

This is definitely not best practice. Aaron and Jack and I have been talking recently, so hopefully the whole PR flow will move better and faster and we can avoid many open inter-related PRs, rather do them in proper order.

Reply:
@alexanderbez you are the only open "changes requested". 

Can you modify that or point out which changes I still need to make?